third-culture-software/bhima
# Add showing lot usage in lots registry and articles in stock

Adds a new function in the action menu for the Stock > Lots registry and the Stock > Articles in Stock registry.
The new action item will display a modal with the suggested order to expend lots for all lots of the same stock article.
The lot usage shows tabular information about each lot in the sequence they should be used.  It also displays a timeline showing part of the same information graphically.

Note that the placement of the lot usage rectangles in the lower, graphical part of the display is based on treating all months as if they were of equal length (30.5 days).  Therefore there may be some very small errors in positioning and width of the items.  For 30/31 day months, the error will be less than one pixel.  But for February it will be more, but probably not very notiecable.

#### TESTING
- Suggest using IMCK for inititial tests and other databases for further testing
- In the Stock > Lots registry 
  - Try the new action menu item (last menu item)
  - Note that the lot you used for the action menu will be shown in light blue
  - Hover over one of the lot usage rectangles in the lower, graphical display to see a message about the lot usage.
  - Note that the initial quantity is shown in black on the left end of the lot rectangle and the red number on the right end of the lot rectangle is what quantity will be left when the lot expires.
  - Note the suggested reorder date
  - Try a lot that has several lots with the same label to see the waterfall/gantt chart display
- In the Stock > Articles in Stock
  - try the same function to see a similar display (no lots will be in blue since this function invoked from an inventory item, not a specific lot).
  - Try it with an inventory item that is all expired (you will need to enable that via the Default search filter).
- Please carefully review the French translations and suggest improvements

For the IMCK database, here are a few Inventory codes that produce displays that show most of the features of this PR:
- Normal inventory items:
    - 002002 Artesuna...
    - 100008 Amoxy...
    - 100022 Ceflacor...
-   Inventory items with CMM=0
    - 100013 Asprin..
    - 100044 Doxy...
    - 100001 Acetazol..

This PR includes various improvements based on discussions with J. Niles.
